src/openms/include/OpenMS/METADATA/ID/QueryMatchGroup.h

```diff
@@ -0,0 +1,96 @@
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Hendrik Weisser $
+// $Authors: Hendrik Weisser $
+// --------------------------------------------------------------------------
+
+#ifndef OPENMS_METADATA_ID_QUERYMATCHGROUP_H
+#define OPENMS_METADATA_ID_QUERYMATCHGROUP_H
+
+#include <OpenMS/METADATA/ID/MoleculeQueryMatch.h>
+
+#include <boost/multi_index_container.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+
+namespace OpenMS
+{
+  namespace IdentificationDataInternal
+  {
+    /*!
+      Group of related (co-identified) molecule-query matches
+
+      E.g. for cross-linking data or multiplexed spectra.
+    */
+    struct QueryMatchGroup: public ScoredProcessingResult
+    {
+      std::set<QueryMatchRef> query_match_refs;
+
+      bool allSameMolecule() const
+      {
+        // @TODO: return true or false for the empty set?
+        if (query_match_refs.size() <= 1) return true;
+        IdentifiedMoleculeRef ref =
+          (*query_match_refs.begin())->identified_molecule_ref;
+        for (auto it = ++query_match_refs.begin(); it != query_match_refs.end();
+             ++it)
+        {
+          if ((*it)->identified_molecule_ref != ref) return false;
+        }
+        return true;
+      }
+
+      bool allSameQuery() const
+      {
+        // @TODO: return true or false for the empty set?
+        if (query_match_refs.size() <= 1) return true;
+        DataQueryRef ref = (*query_match_refs.begin())->data_query_ref;
+        for (auto it = ++query_match_refs.begin(); it != query_match_refs.end();
+             ++it)
+        {
+          if ((*it)->data_query_ref != ref) return false;
+        }
+        return true;
+      }
+    };
+
+    typedef boost::multi_index_container<
+      QueryMatchGroup,
+      boost::multi_index::indexed_by<
+        boost::multi_index::ordered_unique<
+          boost::multi_index::member<QueryMatchGroup, std::set<QueryMatchRef>,
+                                     &QueryMatchGroup::query_match_refs>>>
+      > QueryMatchGroups;
+    typedef IteratorWrapper<QueryMatchGroups::iterator> MatchGroupRef;
```

Comment:
to me this is part of an implementation detail and should not be part of the public API. Also these containers will be hard to access in Python, therefore we should have a public interface that provides access to the objects and the implementation should be private and hidden.

Reply:
Not sure if I understand you correctly. The `QueryMatchGroups` (in this example) needs to be part of the public interface, but the fact that `QueryMatchGroups` is implemented as a multi-index container should be hidden - which I think it is.

Reply:
> Not sure if I understand you correctly. The `QueryMatchGroups` (in this example) needs to be part of the public interface, but the fact that `QueryMatchGroups` is implemented as a multi-index container should be hidden 

I agree

>which I think it is.

if you expose a public typdef like this and use it in other classes, then it becomes part of the public interface, does it not?

Reply:
Let's discuss these interface questions in our upcoming call.